src/borrow.rs

```diff
@@ -129,25 +144,28 @@
 //!
 //! # Limitations
 //!
-//! Note that the current implementation of this is an over-approximation: It will consider all borrows potentially conflicting
-//! if the initial arrays have the same object at the end of their [base object chain][base].
-//! For example, creating two views of the same underlying array by slicing will always yield potentially conflicting borrows
-//! even if the slice indices are chosen so that the two views do not actually share any elements by splitting the array into
-//! non-overlapping parts of by interleaving along one of its axes.
+//! Note that the current implementation of this is an over-approximation: It will consider borrows
+//! potentially conflicting if the initial arrays have the same object at the end of their [base object chain][base].
+//! Then, multiple conditions which are sufficient but not necessary to show the absence of conflicts are checked,
+//! but there are cases which they do not handle, for example slicing an array with a step size
+//! that does not divide its dimension along that axis. In these situations, borrows are rejected even though the arrays
+//! do not actually share any elements.
 //!
 //! This does limit the set of programs that can be written using safe Rust in way similar to rustc itself
 //! which ensures that all accepted programs are memory safe but does not necessarily accept all memory safe programs.
-//! The plan is to refine this checking to correctly handle more involved cases like non-overlapping and interleaved
-//! views into the same array and until then the unsafe method [`PyArray::as_array_mut`] can be used as an escape hatch.
+//! In the future, more involved cases like the example from above may be handled and until then,
+//! the unsafe method [`PyArray::as_array_mut`] can be used as an escape hatch.
```

Comment:
I'm not sure more complex cases will be eventually supported, so I'd rather weaken words here
```suggestion
//! The unsafe method [`PyArray::as_array_mut`] can be used as an escape hatch.
//! More involved cases may be supported in the future, but we don't have a plan yet.
```

Reply:
> but we don't have a plan yet.

But I do have a plan. ;-)

Fully solving the linear Diophantine equation (by recursively solving the two-variable case) for a particular solution and using that, finding the bounds in which all solutions lies should be possible. It is just quite a bit of work, both computationally and w.r.t. the effort to implement it.

What do you think about changing the emphasis, i.e. talk about the escape hatch first and then write

```
//! The unsafe method [`PyArray::as_array_mut`] can be used as an escape hatch.
//! More involved cases like the example from above may be supported in the future.
```

so that there is no "and until then" implying that we will definitely do this eventually?

Reply:
Amended as proposed above.

Reply:
I'm not sure this is such an easy job considering we have different (pointer) ranges for two arrays, but if you are willing to implement it, let's go ahead.

Reply:
> but if you are willing to implement it, let's go ahead slightly_smiling_face

I think it is clear what we would need to do. Of course, actually doing it is not that straight forward.

But my main reason for wanting to release this variant here first and see if there are problems is that doing would probably have a measurable performance impact: We would need to the full strides and dimensions inside `BorrowKey` either adding a dynamic allocation or goring it by `2*32*8=512` bytes. And we would need to do significantly more computations for each borrow. (The code being more complex and more difficult to maintain would not help either.)

So I would say we run with the current version and see were this gets us.

Reply:
Agreed 👍🏼